### ISSUE TITLE
Custom template results in huge difference with alphafold

### ISSUE THREAD

**Issue by empyriumz** [aqlaboratory/openfold#95]
Hi there,

Thanks a lot for your effort to implement trainable AlphaFold in PyTorch.

I came across an interesting [paper](https://doi.org/10.1101/2022.01.07.475350) claiming using templates built with the information from experimental cryo-EM density maps can improve the AlphaFold accuracy.

The authors provide a Colab notebook [here](https://colab.research.google.com/github/phenix-project/Colabs/blob/main/alphafold2/AlphaFold2.ipynb). I tried the notebook, and it worked as intended. 

As an example, the PDB entry 7KU7:
Input fasta sequence: `PLREAKDLHTALHIGPRALSKACNISMQQAREVVQTCPHCNSAPALEAGVNPRGLGPLQIWQTDFTLEPRMAPRSWLAVTVDTASSAIVVTQHGRVTSVAVQHHWATAIAVLGRPKAIKTDNGSCFTSKSTREWLARWGIAHTTGIPGNSQGQAMVERANRLLKDKIRVLAEGDGFMKRIPTSKQGELLAKAMYALNHFERGENTKTPIQKHWRPTVLTEGPPVKIRIETGEWEKGWNVLVWGRGYAAVKNRDTDKVIWVPSRKVKPDITQKDEVTKK`

I supplemented a custom template in CIF format:
https://drive.google.com/file/d/1DUN793nHr0aRRSp29_FwgTGUREwTHcfp/view?usp=sharing

By using this template and turning off the MSA (skip_all_msa == True, equivalent to using dummy MSA), the mean plddt score is about **90**, which is higher than the case with MSA but no custom template.

***

When I tried to replicate the above procedure in OpenFold, however, it looked like the template didn't help. The mean plddt score was less than **40** for model_1 to 5.

To quickly reproduce the results, 

1. I make an empty directory as the path for the `use_precomputed_alignments`, which will lead the data pipeline to use the dummy MSA and an empty template. 

1. Then I load template features generated in the Colab notebook `template_feature_7ku7.pkl` (https://drive.google.com/file/d/1pnZ8pwQZTgcOsHTikQ6X7PQ1bqQs3tqt/view?usp=sharing)

```
import pickle
with open("template_feature_7ku7.pkl", "rb") as f:
    template_feature = pickle.load(f)
feature_dict = {**feature_dict, **template_feature}
```
The rest of the codes are left intact.
So, could you help me check if there is anything wrong with my approach, or is it due to something buggy with template associated codes within the OpenFold?
Thank you very much.

**Comment by gahdritz** [comment]
Could you share your final OpenFold script?

**Comment by gahdritz** [comment]
Great thanks --- I was editing my question just as you responded. Did you use the official DeepMind AlphaFold to get those high pLDDT values or this third-party Colab?

**Comment by empyriumz** [comment]
> 

Oh I use this colab notebook: 
https://colab.research.google.com/github/phenix-project/Colabs/blob/main/alphafold2/AlphaFold2.ipynb
which says 

> This notebook is derived from [ColabFold](https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb) and the DeepMind [AlphaFold2 Colab](https://colab.research.google.com/github/deepmind/alphafold/blob/main/notebooks/AlphaFold.ipynb).

I'm not sure how it's different with the official DeepMind AlphaFold. At least the template processing part is a bit different, that's why I dumped their processed template features and use it in OpenFold

**Comment by gahdritz** [comment]
If you have time, could you try the same hack in the official DeepMind Colab? I'd do it myself but I'm rate limited on Colab ATM.

**Comment by empyriumz** [comment]
> 

No problem, I'll see how the DeepMind Colab perform in this case.

**Comment by empyriumz** [comment]
Hi @gahdritz 
I just finished experiment on official AlphaFold Colab notebook by manually uploading all the input feature dictionary (dummy MSA + custom template).

The mean plddt was above 90 for the first 2 models, then I was cut off due to time limit.

**Comment by gahdritz** [comment]
OK I'll investigate this further.

**Comment by empyriumz** [comment]
Thank you! Let me know if you need more information.

**Comment by gahdritz** [comment]
Thanks for bringing this to our attention. There were ultimately several issues responsible for this discrepancy. First, there were a couple of bugs in the OpenFold template processing pipeline, which I've fixed in 591d10d. Second, OpenFold and AlphaFold differ slightly in the naming of the template atom mask feature. AlphaFold calls it `template_all_atom_masks`, while OpenFold drops the "s" and calls it `template_all_atom_mask`. As bad luck would have it, the pickled template feature dictionary contains one of each, which obscures the problem.

After pulling the latest OpenFold commit and running:
```
template_feature["template_all_atom_mask"] = template_feature["template_all_atom_masks"]
```
repeating your experiment gives an average pLDDT of ~91.21, almost identical to AlphaFold's.

**Comment by empyriumz** [comment]
Thank you so much @gahdritz !
I'll pull the new version and test it myself.
